bootstrap/eks/internal/userdata/node.go

```diff
@@ -138,3 +221,88 @@ func NewNode(input *NodeInput) ([]byte, error) {
 
 	return out.Bytes(), nil
 }
+
+// generateAL2023UserData generates userdata for Amazon Linux 2023 nodes.
+func generateAL2023UserData(input *NodeInput) ([]byte, error) {
+	if err := validateAL2023Input(input); err != nil {
+		return nil, err
+	}
+
+	var buf bytes.Buffer
+
+	// Write MIME header
+	if _, err := buf.WriteString(fmt.Sprintf("MIME-Version: 1.0\nContent-Type: multipart/mixed; boundary=%q\n\n", input.Boundary)); err != nil {
+		return nil, fmt.Errorf("failed to write MIME header: %v", err)
+	}
+
+	// Write shell script part if needed
+	if len(input.PreBootstrapCommands) > 0 || len(input.PostBootstrapCommands) > 0 {
+		shellScriptTemplate := template.Must(template.New("shell").Parse(shellScriptPartTemplate))
+		if err := shellScriptTemplate.Execute(&buf, input); err != nil {
+			return nil, fmt.Errorf("failed to execute shell script template: %v", err)
+		}
+		if _, err := buf.WriteString("\n"); err != nil {
+			return nil, fmt.Errorf("failed to write newline: %v", err)
+		}
+	}
+
+	// Write node config part
+	nodeConfigTemplate := template.Must(template.New("node").Parse(nodeConfigPartTemplate))
+	if err := nodeConfigTemplate.Execute(&buf, input); err != nil {
+		return nil, fmt.Errorf("failed to execute node config template: %v", err)
+	}
+
+	return buf.Bytes(), nil
+}
+
+// getCapacityTypeString returns the string representation of the capacity type.
+func (ni *NodeInput) getCapacityTypeString() string {
+	if ni.CapacityType == nil {
+		return "ON_DEMAND"
+	}
+	switch *ni.CapacityType {
+	case v1beta2.ManagedMachinePoolCapacityTypeSpot:
+		return "SPOT"
+	case v1beta2.ManagedMachinePoolCapacityTypeOnDemand:
+		return "ON_DEMAND"
+	default:
+		return strings.ToUpper(string(*ni.CapacityType))
+	}
+}
+
+// validateAL2023Input validates the input for AL2023 user data generation.
+func validateAL2023Input(input *NodeInput) error {
+	if input.APIServerEndpoint == "" {
+		return fmt.Errorf("API server endpoint is required for AL2023")
+	}
+	if input.CACert == "" {
+		return fmt.Errorf("CA certificate is required for AL2023")
+	}
+	if input.ClusterName == "" {
+		return fmt.Errorf("cluster name is required for AL2023")
+	}
+	if input.NodeGroupName == "" {
+		return fmt.Errorf("node group name is required for AL2023")
+	}
+
+	if input.MaxPods == nil {
+		if input.UseMaxPods != nil && *input.UseMaxPods {
+			input.MaxPods = ptr.To[int32](58)
+		} else {
+			input.MaxPods = ptr.To[int32](110)
+		}
+	}
+	if input.DNSClusterIP == nil {
```

Comment:
DNSCluster IP is an override. So it is quite optional. It is not mandatory to specify.